## Final Project Submission

Please fill out:
* Student name: Matthew Parker
* Student pace: full time
* Scheduled project review date/time: ????
* Instructor name: Rafael Carrasco
* Blog post URL: ????


Importing and Inspecting:

In [2]:
#let's just go ahead and import every python library in existence....just in case  :-)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import statsmodels.api as sm
import statsmodels.stats.api as sms
import statsmodels.formula.api as smf
import scipy.stats as stats
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import cross_val_score

In [3]:
#reading the .csv file into a pandas dataframe
df = pd.read_csv('kc_house_data.csv')

In [4]:
df.head()

id        date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  10/13/2014  221900.0         3       1.00         1180   
1  6414100192   12/9/2014  538000.0         3       2.25         2570   
2  5631500400   2/25/2015  180000.0         2       1.00          770   
3  2487200875   12/9/2014  604000.0         4       3.00         1960   
4  1954400510   2/18/2015  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
0      5650     1.0         NaN   0.0     ...          7        1180   
1      7242     2.0         0.0   0.0     ...          7        2170   
2     10000     1.0         0.0   0.0     ...          6         770   
3      5000     1.0         0.0   0.0     ...          7        1050   
4      8080     1.0         0.0   0.0     ...          8        1680   

   sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0            0.0     1955           0.0    98178  47.5112 -122.257   
1          400.0     1951        1991.0    98125  47.7210 -122.319   
2            0.0     1933           NaN    98028  47.7379 -122.233   
3          910.0     1965           0.0    98136  47.5208 -122.393   
4            0.0     1987           0.0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


Convert the values in the *date* column to the *datetime* datatype.

In [6]:
df.date = pd.to_datetime(df.date)

Check all columns for placeholders, duplicates, or other atypical values.

In [7]:
for col in df.columns:
    print(col, '\n', df[col].value_counts(normalize=True).head(), '\n\n')

id 
 795000620     0.000139
1825069031    0.000093
2019200220    0.000093
7129304540    0.000093
1781500435    0.000093
Name: id, dtype: float64 


date 
 2014-06-23    0.006575
2014-06-25    0.006066
2014-06-26    0.006066
2014-07-08    0.005880
2015-04-27    0.005834
Name: date, dtype: float64 


price 
 350000.0    0.007964
450000.0    0.007964
550000.0    0.007362
500000.0    0.007038
425000.0    0.006945
Name: price, dtype: float64 


bedrooms 
 3    0.454878
4    0.318655
2    0.127796
5    0.074131
6    0.012594
Name: bedrooms, dtype: float64 


bathrooms 
 2.50    0.248970
1.00    0.178312
1.75    0.141131
2.25    0.094782
2.00    0.089364
Name: bathrooms, dtype: float64 


sqft_living 
 1300    0.006390
1400    0.006251
1440    0.006158
1660    0.005973
1010    0.005973
Name: sqft_living, dtype: float64 


sqft_lot 
 5000    0.016576
6000    0.013428
4000    0.011622
7200    0.010187
7500    0.005510
Name: sqft_lot, dtype: float64 


floors 
 1.0    0.494189
2.0    0.381303
1.

According to the reports above, we have several odd things going on:
1. There is a duplicate row in the "id" column at "795000620"
2. Over half of the "sqft_basement" column is "0.0"; additionally, 21% of all values are "?" placeholders; furthermore, the column's datatype is string instead of numeric
3. The "waterfront" column is over 99% "0.0"; this column is also missing 2,376 values (~11% of total)
4. The vast majority (90%) of values in the "view" column are "0.0"; this is in addition to 63 missing values
5. The "yr_renovated" column is 95.8% "0.0"; this column is also missing 3,842 values (~17.8% of total)

Let's deal with these one at a time, beginning with the duplicate instance of "795000620". Let's display both instances to make sure that they contain the same data.

In [8]:
df.loc[df['id']==795000620]

id       date     price  bedrooms  bathrooms  sqft_living  \
17588  795000620 2014-09-24  115000.0         3        1.0         1080   
17589  795000620 2014-12-15  124000.0         3        1.0         1080   
17590  795000620 2015-03-11  157000.0         3        1.0         1080   

       sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
17588      6250     1.0         0.0   0.0     ...          5        1080   
17589      6250     1.0         0.0   0.0     ...          5        1080   
17590      6250     1.0         NaN   0.0     ...          5        1080   

       sqft_basement yr_built  yr_renovated  zipcode      lat    long  \
17588            0.0     1950           0.0    98168  47.5045 -122.33   
17589            0.0     1950           0.0    98168  47.5045 -122.33   
17590            0.0     1950           NaN    98168  47.5045 -122.33   

       sqft_living15  sqft_lot15  
17588           1070        6250  
17589           1070        6250  
17590           1070        6250  

[3 rows x 21 columns]

So it turns out that this house was sold several times over the span of a few months. This is fine, but notice that some of the columns have missing values (like 'waterfront' and 'yr_renovated'). We should probably check to make sure that the rest of the dataset doesn't have any instances of duplicates where missing values can be filled in.

In [9]:
# since we're about to slice out scattered rows, we should create a 
# new column that duplicates the original index. This will make
# merging the sliced data back in much easier later on.
df = df.reset_index()

In [10]:
duplicates = df[df.duplicated(subset='id', keep=False)]

# How many duplicates do we have?
print(len(duplicates))

353


In [11]:
# let's see if there are any houses which *only* have NaN values for 'waterfront'
duplicates.loc[duplicates['waterfront']!=0].sort_values('id')

index          id       date      price  bedrooms  bathrooms  \
2494    2494     1000102 2014-09-16   280000.0         6       3.00   
824      824   726049190 2015-02-18   431000.0         3       1.00   
17590  17590   795000620 2015-03-11   157000.0         3       1.00   
10261  10261  1450100390 2014-09-05   125000.0         3       1.00   
4868    4868  1524079093 2015-03-18   369500.0         3       1.75   
5718    5718  1788800630 2014-10-29    96500.0         3       1.00   
3536    3536  1823049202 2014-06-10   175000.0         6       1.50   
3947    3947  1825069031 2014-10-16   550000.0         4       1.75   
19195  19195  1901600090 2015-04-26   390000.0         5       1.75   
1862    1862  2143700830 2014-10-06   207000.0         4       2.50   
9384    9384  2212200100 2015-04-22   344900.0         4       2.50   
9383    9383  2212200100 2014-10-22   229950.0         4       2.50   
1449    1449  2228900270 2015-02-12   302000.0         2       1.00   
371      371  2231500030 2014-10-01   315000.0         4       2.25   
15555  15555  2560801222 2014-11-13   309950.0         3       2.25   
15986  15986  2561340020 2014-08-04   325000.0         3       1.75   
9711    9711  2568300045 2014-06-25   305000.0         6       2.00   
20596  20596  2724049222 2014-08-02   163800.0         2       2.50   
5587    5587  2726049071 2014-12-11   510000.0         2       1.00   
11764  11764  2892700041 2015-01-28   238000.0         3       1.50   
13605  13605  3185600040 2014-12-24   310000.0         2       1.00   
10211  10211  3293700496 2014-12-04   450000.0         4       1.75   
5240    5240  3303000130 2015-04-28   520000.0         3       2.25   
18748  18748  3332000615 2015-04-22   389000.0         3       1.00   
5268    5268  4204400098 2015-04-21   385000.0         5       1.75   
11446  11446  4305200070 2014-05-19   350000.0         3       2.25   
2125    2125  4345000510 2015-04-28   325000.0         3       2.50   
9270    9270  4443800385 2014-08-18   410000.0         2       1.00   
9480    9480  5054800110 2014-10-16   238000.0         5       2.25   
9481    9481  5054800110 2015-02-13   328000.0         5       2.25   
8524    8524  5101405604 2015-04-28   395000.0         1       1.00   
8622    8622  5127001320 2015-02-23   314950.0         3       1.75   
12327  12327  5132000140 2015-01-20   415000.0         6       1.00   
6896    6896  5249801440 2014-12-16   250000.0         3       1.00   
6897    6897  5249801440 2015-04-22   570000.0         3       1.00   
6713    6713  5417600130 2014-10-10   244500.0         2       1.00   
11049  11049  6141100320 2014-07-07   245000.0         2       1.00   
13285  13285  7167000040 2014-08-13   740000.0         4       3.00   
14988  14988  7504021310 2014-12-04   745000.0         3       2.50   
324      324  7520000520 2014-09-05   232000.0         2       1.00   
2563    2563  7856400240 2015-02-11  1650000.0         4       3.00   
20654  20654  8564860270 2015-03-30   502000.0         4       2.50   
17052  17052  8651402750 2015-02-18   132825.0         3       1.50   
11574  11574  8807810890 2014-08-27   259875.0         3       1.00   
2974    2974  9136103130 2014-12-01   430000.0         2       1.50   
6365    6365  9222400605 2014-11-15   842500.0         5       4.00   

       sqft_living  sqft_lot  floors  waterfront     ...      grade  \
2494          2400      9373     2.0         NaN     ...          7   
824           1810      7200     1.0         NaN     ...          7   
17590         1080      6250     1.0         NaN     ...          5   
10261          920      7314     1.0         NaN     ...          6   
4868          1300     20700     1.0         NaN     ...          7   
5718           840     12091     1.0         NaN     ...          6   
3536          1930      8400     1.0         NaN     ...          7   
3947          2410      8447     2.0         NaN     ...          8   
19195         1940      6654  

In [12]:
# let's see if there are any houses which *only* have NaN values for 'yr_renovated'
duplicates.loc[duplicates['yr_renovated']!=0].sort_values('id')

index          id       date      price  bedrooms  bathrooms  \
16801  16801     7200179 2015-04-24   175000.0         2       1.00   
12406  12406   123039336 2014-12-08   244900.0         1       1.00   
14841  14841   324000530 2014-07-08   201500.0         3       1.00   
17590  17590   795000620 2015-03-11   157000.0         3       1.00   
12422  12422  1217000340 2014-06-06   185000.0         3       1.00   
8497    8497  1237500540 2014-12-22   270000.0         3       1.75   
3752    3752  1254200015 2014-12-16   405000.0         3       2.50   
14560  14560  1432400120 2014-11-11   165000.0         3       1.00   
10262  10262  1450100390 2015-03-16   208000.0         3       1.00   
4867    4867  1524079093 2014-08-27   275000.0         3       1.75   
8427    8427  1721801010 2015-04-24   302100.0         3       1.00   
8426    8426  1721801010 2014-09-03   225000.0         3       1.00   
2500    2500  1781500435 2014-08-20   260000.0         3       1.75   
3946    3946  1825069031 2014-08-14   550000.0         4       1.75   
3947    3947  1825069031 2014-10-16   550000.0         4       1.75   
9112    9112  1922059278 2015-03-05   255000.0         3       1.00   
9111    9111  1922059278 2014-10-14   145000.0         3       1.00   
5753    5753  1954420170 2014-05-21   368250.0         3       2.50   
5754    5754  1954420170 2014-11-13   580000.0         3       2.50   
17277  17277  1974300020 2015-02-18   624900.0         4       2.50   
12943  12943  2019200220 2015-02-26   269000.0         3       2.25   
13616  13616  2023049218 2015-03-16   445000.0         2       1.00   
15787  15787  2044500213 2014-06-17   310000.0         4       2.00   
1448    1448  2228900270 2014-08-12   215000.0         2       1.00   
9712    9712  2568300045 2015-03-19   649950.0         6       2.00   
15264  15264  2619920170 2014-12-19   765000.0         4       2.50   
5588    5588  2726049071 2015-04-08   489950.0         2       1.00   
3620    3620  2767602141 2014-12-22   650000.0         3       1.50   
10599  10599  2767603612 2015-01-13   489000.0         2       2.25   
11763  11763  2892700041 2014-07-14   168000.0         3       1.50   
...      ...         ...        ...        ...       ...        ...   
3790    3790  4435000705 2015-03-09   255500.0         3       1.00   
13722  13722  5332200530 2015-04-24  1020000.0         5       2.50   
19321  19321  6021500970 2015-04-07   874950.0         2       1.00   
11049  11049  6141100320 2014-07-07   245000.0         2       1.00   
15142  15142  6669020290 2014-10-23   169000.0         3       1.75   
18220  18220  6751300375 2014-10-16   522500.0         3       1.00   
15796  15796  7129304540 2014-12-20   133000.0         5       2.00   
17469  17469  7387500235 2015-03-17   363000.0         3       1.75   
324      324  7520000520 2014-09-05   232000.0         2       1.00   
325      325  7520000520 2015-03-11   240500.0         2       1.00   
12110  12110  7520000695 2015-04-21   251000.0         3       1.00   
5247    5247  7524400250 2014-11-24   589950.0         3       2.00   
21565  21565  7853420110 2015-05-04   625000.0         3       3.00   
1575    1575  7888000390 2015-04-01   235000.0         3       1.00   
12365  12365  7899800045 2014-08-28   107000.0         3       1.50   
15351  15351  7972000010 2014-05-20   120750.0         3       1.75   
15352  15352  7972000010 2014-10-21   195000.0         3       1.75   
2629    2629  7983000200 2014-10-05   169575.0         3       1.00   
1128    1128  8062900070 2015-02-13   369000.0         5       1.50   
9269    9269  8129700644 2015-04-24   780000.0         3       4.00   
6527    6527  8645530010 2014-05-15   225000.0         3       2.00   
6783    6783  8651510380 2014-08-21   310000.0         3       2.00   
11575  11575  8807810890 2014-11-05   385000.0         3       1.00   
718      718  8820903380 2015-01-02   730000.0         6       2.25   
717      717  8820903380 2014-0

So looking at the duplicated data, it seems that we would be safe to convert all 'waterfront' NaN values to '0.0', except for *id*'s '2212200100', '5054800110', and '5249801440' (which only have NaN values in both instances). The same is true for all 'yr_renovated' NaN values except for *id*'s '1922059278', '1954420170', '3262300940', '4222310010', and '7972000010'.

So, first we will remove the seven pairs of rows just mentioned since we cannot reliably fill in any missing data. Then, let's replace all of the NaN values in 'waterfront' and 'yr_renovated' with '0.0'. Finally, we will merge the *duplicates* DataFrame back into the original *df* DataFrame so that our changes are updated to the original.

In [13]:
# change index to 'id' so that we can tell pandas which rows to remove
duplicates = duplicates.set_index('id')
duplicates.drop([2212200100, 5054800110, 5249801440, 1922059278, 1954420170, 
                 3262300940, 4222310010, 7972000010], inplace=True)

# now reset the index to preserve the 'id' column, then set the index back
# to 'index' for merging with the original DataFrame
duplicates = duplicates.reset_index()
duplicates = duplicates.set_index('index')

In [14]:
# now replace all of the NaN values in 'waterfront' and 'yr_renovated' with '0.0'
duplicates.waterfront.fillna(0, inplace=True)
duplicates.yr_renovated.fillna(0, inplace=True)

In [15]:
# Now let's merge the duplicates DataFrame back into the original.
# first set df's index back to 'index' as it originally was.
df.set_index('index', inplace=True)
df.update(duplicates, overwrite=True)

# because pandas converts all integers to floats in an update, we need to set them back
df = df.astype({'id': 'int64', 'bedrooms': 'int64', 'sqft_living': 'int64', 'sqft_lot': 'int64', 
                  'condition': 'int64', 'grade': 'int64', 'sqft_above': 'int64', 'yr_built': 'int64', 
                  'zipcode': 'int64', 'sqft_living15': 'int64', 'sqft_lot15': 'int64'})

Alright, that's 1 down, only 4 more weird features to go. 

Let's tackle the problematic "sqft_basement" column next. First, we'll turn all the "?" values into NaN values. Then we'll convert the entire column into numeric data using the *.to_numeric* method.

In [16]:
# we define a function that will replace all "?" values with NaN
def impute_sqft_basement(value):
    if value == '?':
        return None
    else:
        return value

# now we map a lamba function to iterate the above "impute" function through
# the *sqft_basement* column
df.sqft_basement = df.sqft_basement.map(lambda x: impute_sqft_basement(x))

In [17]:
# and now we will change the data type to numeric
df.sqft_basement = pd.to_numeric(df.sqft_basement)

Okay, that's another weird column dealt with. 

How about the *waterfront* and *yr_renovated* columns? Since both column are missing over 10% of their data *and* around 95-99% of the known values are identical, it should be safe to exclude these variables from our model. The values for the variable *view* are also 90% identical, so let's drop it as well.

In [20]:
# create a new dataframe labeled as a subset of the original, then use the *.drop* method
df_subset = df.drop(['waterfront', 'yr_renovated', 'view'], axis=1)

OK, now we just have left the missing values in *sqft_basement* to contend with. About 60% of the properties have no basement and there are 454 NaN values. Intuitively, it seems like the presence/absence of a basement is likely to impact the sale price of a home, so instead of dropping the column let's just drop the rows with NaN values instead. This is only 2% of our dataset, so our model should not be significantly effected.

In [25]:
df_subset.dropna(subset=['sqft_basement'], inplace=True)


Now that our data has been sufficiently cleaned, let's get a look at our variables and whether they might be related.


In [ ]:
# let's use a scatter matrix to take a quick look at the variables
pd.scatter_matrix(df_subset, figsize=(20,20));

C:\Users\Matthew Parker\.conda\envs\learn-env\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: pandas.scatter_matrix is deprecated, use pandas.plotting.scatter_matrix instead
  


array([[<matplotlib.axes._subplots.AxesSubplot object at 0x00000191E2FC4550>,
      dtype=object)